std/experimental/allocator/building_blocks/allocator_list.d

```diff
@@ -196,6 +196,57 @@ struct AllocatorList(Factory, BookkeepingAllocator = GCAllocator)
         return null;
     }
 
+    /**
+    Allocate a block of size `s` with alignment `a`. First tries to allocate
+    from the existing list of already-created allocators. If neither can
+    satisfy the request, creates a new allocator by calling `make(s + a - 1)`
+    and delegates the request to it. However, if the allocation fresh off a
+    newly created allocator fails, subsequent calls to `alignedAllocate`
+    will not cause more calls to `make`.
+    */
+    static if (hasMember!(Allocator, "alignedAllocate"))
+    void[] alignedAllocate(size_t s, uint theAlignment)
+    {
+        import std.algorithm.comparison : max;
+        import core.checkedint : adds;
+
+        for (auto p = &root, n = *p; n; p = &n.next, n = *p)
+        {
+            auto result = n.alignedAllocate(s, theAlignment);
+            if (result.length != s) continue;
+            // Bring to front if not already
+            if (root != n)
+            {
+                *p = n.next;
+                n.next = root;
+                root = n;
+            }
+            return result;
+        }
+        // Can't allocate from the current pool. Check if we just added a new
+        // allocator, in that case it won't do any good to add yet another.
+        if (root && root.empty == Ternary.yes)
+        {
+            // no can do
+            return null;
+        }
+
+        bool overflow = false;
+        size_t maxSize = adds(s - 1, cast(size_t) theAlignment, overflow);
+        assert(!overflow, "Requested size is too large");
+        if (overflow)
```

Comment:
Because of the `assert(!overflow)` above, [by the language specification the compiler may assume that `overflow` is never true](https://dlang.org/spec/contracts.html#assert_contracts) (although I do not believe any D compiler currently does this):
>The compiler is free to assume the assert expression is true and optimize subsequent code accordingly.

Reply:
For debug builds, I assume that the assert will not be optimized.
However, in case an overflow happens, I have the check right below the assert which should never be optimized away.